### PR TITLE
V2XHub / CARMA Streets to process and broadcast SSM

### PR DESCRIPTION
**Pull request by dan-du-car**
# PR Details
## Description
For the POC TIM/TSP use case, priority-eligible vehicles will broadcast J2375 Signal Request Message (SRM) to the MMITSS Roadside Processor (MRP), which in turn will send J2375 Signal Status Message (SSM) in acknowledgement.  

Since MRP is being integrated with CARMA Streets, CARMA Streets / V2XHub needs to #1) consume the json SSM (used by the MMITSS MRP internally, see [here](https://github.com/mmitss/mmitss-az/blob/master/src/mrp/priority-request-server/Readme.md) and [here](https://github.com/mmitss/mmitss-az/blob/master/src/mrp/priority-request-server/SSM.json)) sent on the CARMA Streets Kafka broker, #2) encode it according to the J2375 [ASN.1 ](https://leidoscorpus.sharepoint.us/:f:/s/STR/EpU-cLOWhUtGvsddHUSf8I4Bf7-Ot8oMIY4yf2m7x2-Uag?e=0Rvpug)schema, and #3) broadcast J2375 SSMs through RSU.  

For #1) the V2XHub carma-streets-plugin needs to pull the json SSM from CARMA Steets Kafa broker.

For #3) the V2XHub immediate-forward-plugin configuration needs to to be updated.

## Related Issue
https://github.com/usdot-fhwa-OPS/V2X-Hub/issues/534

## Motivation and Context
TM/TSP

## How Has This Been Tested?
Unit test

## Types of changes

- [ ] Defect fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:

- [ ] I have added any new packages to the sonar-scanner.properties file
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
[V2XHUB Contributing Guide](https://github.com/usdot-fhwa-OPS/V2X-Hub/blob/develop/Contributing.md) 
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: src/v2i-hub/CARMAStreetsPlugin/src/JsonToJ2735SSMConverter.cpp [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+            std::cout << "Parse error: " << err << endl;
+        }
+        return parseResult;
+    }
+
+    void JsonToJ2735SSMConverter::toJ2735SSM(const Json::Value &ssmDoc, std::shared_ptr<SignalStatusMessage> ssmPtr) const
+    {
+        if (!ssmDoc.isMember("SignalStatus"))
+        {
+            return;
```
Comment (paulbourelly999) [inline src/v2i-hub/CARMAStreetsPlugin/src/JsonToJ2735SSMConverter.cpp:24]: What happens if we return here without throwing exception. Does the calling method handle this error vase somehow.

**Comment by paulbourelly999** [comment]
@dan-du-car Sonar scan report seems to indicate several potential memory leaks. Can we confirm that we are deallocating memory even in error cases when creating the SSM from JSON

File: src/v2i-hub/CARMAStreetsPlugin/src/CARMAStreetsPlugin.cpp [hunk 2]
```diff
@@ -659,6 +679,71 @@ void CARMAStreetsPlugin::SubscribeSpatKafkaTopic(){
...
+					{
+						ssm_convertor.encodeSSM(ssm_ptr, ssmEncodedMsg);
+					}
+					catch (TmxException &ex) 
+					{
+						// Skip messages that fail to encode.
+						PLOG(logERROR) << "Failed to encoded SSM message : \n" << payload_str << std::endl << "Exception encountered: " 
+							<< ex.what() << std::endl;
+						ASN_STRUCT_FREE_CONTENTS_ONLY(asn_DEF_SignalStatusMessage, ssm_ptr.get());
+						SetStatus<uint>(Key_SPATMessageSkipped, ++_spatMessageSkipped);
```
[the remark this answers is not in the document]
Comment (dan-du-car) [reply src/v2i-hub/CARMAStreetsPlugin/src/CARMAStreetsPlugin.cpp]: updated

**Comment by paulbourelly999** [comment]
> @dan-du-car Sonar scan report seems to indicate several potential memory leaks. Can we confirm that we are deallocating memory even in error cases when creating the SSM from JSON
> 
> @paulbourelly999 I think this is where the deallocating memory when the SSM is not fully populated and SSM is failed to encode:
> 
> https://github.com/usdot-fhwa-OPS/V2X-Hub/blob/050d28a3b8a37c357f25335464d2080522bab7ea/src/v2i-hub/CARMAStreetsPlugin/src/CARMAStreetsPlugin.cpp#L662
> 
> I do not expect throwing fatal exceptions in the converter as I check every members of the JSON to make sure it exists, and data type is correct before it is used to populate the SSM. If the JSON has incorrect data type in those fields, it will skip them.

Isn't this the SPat and not the SSM?

File: src/v2i-hub/CARMAStreetsPlugin/src/JsonToJ2735SSMConverter.cpp [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+                    asn_sequence_add(&signalStatus->sigStatus.list.array, signalStatusPackage);
+                } // Populate signal status package
+            }
+
+            asn_sequence_add(&statusPtr->list.array, signalStatus);
+            ssmPtr->status = *statusPtr;
+        }
+        catch(exception &ex)
+        {
+            std::cout << "ERROR: Cannot read JSON file." << std::endl;
```
Comment (paulbourelly999) [inline src/v2i-hub/CARMAStreetsPlugin/src/JsonToJ2735SSMConverter.cpp:76]: Please correct this to use PLOG macros and this is not a JSON file it is a JSON message.

File: src/v2i-hub/CARMAStreetsPlugin/src/JsonToJ2735SSMConverter.cpp [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+    }
+
+    void JsonToJ2735SSMConverter::toJ2735SSM(const Json::Value &ssmDoc, std::shared_ptr<SignalStatusMessage> ssmPtr) const
+    {
+        try 
+        {        
+            ASN_STRUCT_FREE_CONTENTS_ONLY(asn_DEF_SignalStatusMessage, ssmPtr.get());
+            if (!ssmDoc.isMember("SignalStatus"))
+            {
+                return;
```
Comment (paulbourelly999) [inline src/v2i-hub/CARMAStreetsPlugin/src/JsonToJ2735SSMConverter.cpp:27]: Maybe add a log statement to indicate something has gone wrong here

Comment (dan-du-car) [reply src/v2i-hub/CARMAStreetsPlugin/src/JsonToJ2735SSMConverter.cpp:28]: https://github.com/usdot-fhwa-OPS/V2X-Hub/pull/533/commits/53ec358d4de07c0a7567a9b5dfe7e86e0a5624c3

File: src/v2i-hub/CARMAStreetsPlugin/src/JsonToJ2735SSMConverter.cpp [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+    bool JsonToJ2735SSMConverter::parseJsonString(const string &consumedMsg, Json::Value &ssmDoc) const
+    {
+        const auto jsonLen = static_cast<int>(consumedMsg.length());
+        Json::CharReaderBuilder builder;
+        JSONCPP_STRING err;
+        const std::unique_ptr<Json::CharReader> reader(builder.newCharReader());
+        bool parseResult = reader->parse(consumedMsg.c_str(), consumedMsg.c_str() + jsonLen, &ssmDoc, &err);
+        if (!parseResult)
+        {
+            std::cout << "Parse error: " << err << endl;
```
[the remark this answers is not in the document]
Comment (dan-du-car) [reply src/v2i-hub/CARMAStreetsPlugin/src/JsonToJ2735SSMConverter.cpp]: https://github.com/usdot-fhwa-OPS/V2X-Hub/pull/533/commits/53ec358d4de07c0a7567a9b5dfe7e86e0a5624c3

File: src/v2i-hub/CARMAStreetsPlugin/src/JsonToJ2735SSMConverter.cpp (hunk 3 above)
Comment (dan-du-car) [reply src/v2i-hub/CARMAStreetsPlugin/src/JsonToJ2735SSMConverter.cpp]: https://github.com/usdot-fhwa-OPS/V2X-Hub/pull/533/commits/53ec358d4de07c0a7567a9b5dfe7e86e0a5624c3

**Review by paulbourelly999 — APPROVED** [review]

**Pull request merged**